tests/unittest_nodes_lineno.py

```diff
@@ -0,0 +1,1185 @@
+        assert (n4.lineno, n4.col_offset) == (5, 0)
+        assert (n4.end_lineno, n4.end_col_offset) == (7, 1)
+        assert (n4.targets[0].lineno, n4.targets[0].col_offset) == (5, 0)
+        assert (n4.targets[0].end_lineno, n4.targets[0].end_col_offset) == (5, 4)
+
+    @staticmethod
+    def test_end_lineno_attribute() -> None:
+        """Attribute, AssignAttr, DelAttr."""
+        code = textwrap.dedent(
+            """
+        class X:
+            var = 42
+
+        X.var2 = 2  #@
+        X.var2  #@
+        del X.var2  #@
+        """
+        ).strip()
+        ast_nodes = builder.extract_node(code)
+        assert isinstance(ast_nodes, list) and len(ast_nodes) == 3
+
+        a1 = ast_nodes[0]
+        assert isinstance(a1, nodes.Assign)
+        assert isinstance(a1.targets[0], nodes.AssignAttr)
+        assert isinstance(a1.value, nodes.Const)
+        assert (a1.lineno, a1.col_offset) == (4, 0)
+        assert (a1.end_lineno, a1.end_col_offset) == (4, 10)
+        assert (a1.targets[0].lineno, a1.targets[0].col_offset) == (4, 0)
+        assert (a1.targets[0].end_lineno, a1.targets[0].end_col_offset) == (4, 6)
+        assert (a1.value.lineno, a1.value.col_offset) == (4, 9)
+        assert (a1.value.end_lineno, a1.value.end_col_offset) == (4, 10)
+
+        a2 = ast_nodes[1]
+        assert isinstance(a2, nodes.Attribute) and isinstance(a2.expr, nodes.Name)
+        assert (a2.lineno, a2.col_offset) == (5, 0)
+        assert (a2.end_lineno, a2.end_col_offset) == (5, 6)
+        assert (a2.expr.lineno, a2.expr.col_offset) == (5, 0)
+        assert (a2.expr.end_lineno, a2.expr.end_col_offset) == (5, 1)
+
+        a3 = ast_nodes[2]
+        assert isinstance(a3, nodes.Delete) and isinstance(a3.targets[0], nodes.DelAttr)
+        assert (a3.lineno, a3.col_offset) == (6, 0)
+        assert (a3.end_lineno, a3.end_col_offset) == (6, 10)
+        assert (a3.targets[0].lineno, a3.targets[0].col_offset) == (6, 4)
+        assert (a3.targets[0].end_lineno, a3.targets[0].end_col_offset) == (6, 10)
+
+    @staticmethod
+    def test_end_lineno_call() -> None:
+        """Call, Keyword."""
+        code = textwrap.dedent(
+            """
+        func(arg1, arg2=value)  #@
+        """
+        ).strip()
+        c1 = builder.extract_node(code)
+        assert isinstance(c1, nodes.Call)
+        assert isinstance(c1.func, nodes.Name)
+        assert isinstance(c1.args[0], nodes.Name)
+        assert isinstance(c1.keywords[0], nodes.Keyword)
+        assert isinstance(c1.keywords[0].value, nodes.Name)
+
+        assert (c1.lineno, c1.col_offset) == (1, 0)
+        assert (c1.end_lineno, c1.end_col_offset) == (1, 22)
+        assert (c1.func.lineno, c1.func.col_offset) == (1, 0)
+        assert (c1.func.end_lineno, c1.func.end_col_offset) == (1, 4)
+
+        assert (c1.args[0].lineno, c1.args[0].col_offset) == (1, 5)
+        assert (c1.args[0].end_lineno, c1.args[0].end_col_offset) == (1, 9)
+
+        if PY39_PLUS:
```

Comment:
Should we test that these are `None` for < `3.9`?

Reply:
I'll add `else` blocks. If we manually fix those issue for Python 3.8 in ` TreeRebuilder` these should let us know.